File: packages/shared/src/types/snapshot.ts
/**
 * AssetSnapshot（总资产每日唯一记录）类型 —— **归并转发模块**（T01 共享类型去重）
 *
 * 【为什么只剩转发】
 * `AssetSnapshot` 曾在 `packages/shared/src/types.ts`（包入口 `src/index.ts`
 * re-export 的那份）与本文件各写一份。本次统一以 `../types.ts` 为**唯一真相源**，
 * 本文件退化为 type-only 转发，`src/types/index.ts` 的
 * `export * from './snapshot.ts'` 行为完全不变。
 *
 * 【语义备忘（原文档保留）】
 * 每个组合每天最多一条记录（唯一约束 portfolioId + date）；
 * source = DERIVED（系统按交易/余额推导）/ MANUAL（用户手工录入，取代当日自动值）；
 * valuationFlag 标注估值口径（正常 / 沿用前值 / 按成本估值）。
 *
 * ⚠️ 请勿在此新增声明；新增/修改一律去 `packages/shared/src/types.ts`。
 */

export type {
  /** 总资产每日唯一记录实体（对应 Prisma model AssetSnapshot） */
  AssetSnapshot,
} from '../types.ts';

File: backend/app/models/snapshot.py
"""总资产每日唯一记录表（对齐 app Prisma: AssetSnapshot）。

派生层 + 手工录入；唯一键 (portfolio_id, date)。recorded_at 为记录写入时间，
与 created_at/updated_at 并列（Prisma 三列独立）。
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Index,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, pk_uuid
from app.models.enums import SnapshotSource, SnapshotValuation


class AssetSnapshot(Base, TimestampMixin):
    __tablename__ = "asset_snapshots"
    __table_args__ = (
        UniqueConstraint("portfolio_id", "date", name="uq_asset_snapshots_portfolio_date"),
        Index("ix_asset_snapshots_portfolio_date", "portfolio_id", "date"),
    )

    id: Mapped[str] = pk_uuid()
    portfolio_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("portfolios.id", ondelete="CASCADE"),
        nullable=False,
    )
    date: Mapped[date] = mapped_column(Date, nullable=False)
    total_asset: Mapped[Decimal] = mapped_column(
        "total_asset", Numeric(18, 2), nullable=False
    )
    market_value: Mapped[Decimal | None] = mapped_column(
        "market_value", Numeric(18, 2), nullable=True
    )
    cash_balance: Mapped[Decimal | None] = mapped_column(
        "cash_balance", Numeric(18, 2), nullable=True
    )
    source: Mapped[SnapshotSource] = mapped_column(
        Enum(SnapshotSource, name="SnapshotSource", native_enum=True, create_type=True),
        nullable=False,
    )
    valuation_flag: Mapped[SnapshotValuation] = mapped_column(
        "valuation_flag",
        Enum(SnapshotValuation, name="SnapshotValuation", native_enum=True, create_type=True),
        nullable=False,
    )
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    recorded_at: Mapped[datetime] = mapped_column(
        "recorded_at",
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    portfolio: Mapped["Portfolio"] = relationship(back_populates="snapshots")
